sympy/core/add.py

```diff
@@ -835,10 +833,29 @@ def _eval_as_leading_term(self, x):
             return expr.as_leading_term(x)
 
         infinite = [t for t in expr.args if t.is_infinite]
+        compute = False
+
+        leading_term = [t.as_leading_term(x) for t in expr.args]
+        new_expr = expr.func(*leading_term).removeO()
+
+        if not new_expr:
+            compute = True
+        elif new_expr is not S.NaN:
+            if new_expr.is_Add:
+                final_leading_term = [t for t in new_expr.args]
+            else:
+                final_leading_term = [new_expr]
+
+            if len(expr.args) != len(final_leading_term):
+                canceled_term = [t for t in leading_term if t not in final_leading_term]
+                expr_sum = expr.func(*canceled_term)
+                if expr_sum == S(0):
```

Comment:
Can you give an example where expr_sum is different from 0?

Reply:
if ```leading_term``` is like ```[sin(a)**2,cos(a)**2] ``` then new_expr will be 1. In that case expr_sum will be 1. As canceled_term is [sin(a)**2,cos(a)**2]

Reply:
if leading_term = [3*x,5*x,2*x] then new_expr will be 8*x.canceled_term = [3*x,5*x,2*x] And expr_sum is 10*x.
consider ```expr = 3*x+sin(5*x)+2*x/(1-x)```

Reply:
```
>>> self = 3*x + 2*x/(-x + 1) + sin(5*x)
>>> expr=expand_mul(self)
>>> leading_term = [t.as_leading_term(x) for t in expr.args]
>>> new_expr = expr.func(*leading_term).removeO()
>>> if new_expr.is_Add:
        final_leading_term = [t for t in new_expr.args]
else:
        final_leading_term = [new_expr]

>>> canceled_term = [t for t in leading_term if t not in final_leading_term]
>>> final_leading_term
[10*x]
>>> canceled_term
[3*x, 2*x, 5*x]
>>> expr_sum = expr.func(*canceled_term)
>>> expr_sum
10*x
```

Reply:
It seems that the condition `expr_sum == 0` does not work in this example:
```
>>> e = 1/(1 - x) - 1 + sin(x) + x
>>> e.as_leading_term(x)
2*x
```